### DISCUSSION TITLE
How to enable nvim-treesitter git instead of curl

### DISCUSSION THREAD

**Discussion by emcdonald** [LazyVim/LazyVim#737] [Q&A] [answered]
I'm trying to [enable nvim-treesitter's prefer_git](https://github.com/nvim-treesitter/nvim-treesitter#i-want-to-use-git-instead-of-curl-for-downloading-the-parsers) setting. Typically this is done with the following in a lua config:

`require("nvim-treesitter.install").prefer_git = true`

I'm having trouble determining how this can be done.

Any help is much appreciated!

**Comment by dpetka2001** [comment]
Something like
```lua
return {
  {
    "nvim-treesitter/nvim-treesitter",
    opts = {
      prefer_git = true,
    },
  },
}
```
should work I believe. But I don't know how to properly find evidence that it works. Is there some kind of treesitter log somewhere? It shows in `TSConfigInfo` that it should be set.

**Comment by emcdonald** [comment]
Below is the output of my TSConfigInfo after using your suggestion. As you can see, `prefer_git` is nested inside of `modules`. Treesitter is still trying to use curl. I haven't been able to find a log that Treesitter is producing, either.

I'm not sure where the TSConfig expects `prefer_git` to live. After a look at the nvim-treesitter source code, `prefer_git` is held in its [`install` module](https://github.com/nvim-treesitter/nvim-treesitter/blob/6e8ad92031ff3b07db8980ab4a5491d1c40398ea/lua/nvim-treesitter/install.lua#L20).

Meanwhile, the config items that Lazy is tweaking seem to all live in the [`configs` module in the TSConfig structure](https://github.com/nvim-treesitter/nvim-treesitter/blob/6e8ad92031ff3b07db8980ab4a5491d1c40398ea/lua/nvim-treesitter/configs.lua#L20).

Thank you for the help.

```lua
{
  auto_install = false,
  ensure_installed = { "bash", "c", "html", "javascript", "json", "lua", "luadoc", "luap", "markdown", "markdown_inline", "python", "query", "regex", "tsx", "typescript", "vim", "vimdoc", "yaml" },
  ignore_install = {},
  modules = {
    context_commentstring = {
      enable = true,
      enable_autocmd = false
    },
    highlight = {
      additional_vim_regex_highlighting = false,
      custom_captures = {},
      disable = {},
      enable = true,
      loaded = true,
      module_path = "nvim-treesitter.highlight"
    },
    illuminate = {
      disable = {},
      enable = true,
      loaded = true,
      module_path = "illuminate.providers.treesitter"
    },
    incremental_selection = {
      disable = {},
      enable = true,
      keymaps = {
        init_selection = "<C-space>",
        node_decremental = "<bs>",
        node_incremental = "<C-space>",
        scope_incremental = false
      },
      loaded = true,
      module_path = "nvim-treesitter.incremental_selection"
    },
    indent = {
      disable = {},
      enable = true,
      loaded = true,
      module_path = "nvim-treesitter.indent"
    },
    prefer_git = true
  },
  sync_install = false
}
```

**Reply by amaanq** [reply]
you put prefer_git inside the modules table, that's the issue

**Reply by emcdonald** [reply]
Do you know how to not automatically insert `prefer_git` into the `modules` table? If I change a property like `auto_install`, it overwrites the old value correctly. However, a "new" value like `prefer_git` that does not exist in `TSConfig` is always inserted into the `modules` table.

Also, even if `prefer_git` appeared in the top level of the TSConfig table, I'm not sure treesitter would load it from there, since `prefer_git` is held in the `install` module and not referenced by nvim-treesitter from the TSConfig table.

**Reply by dpetka2001** [reply]
You should show your config. My example above didn't put `prefer_git` inside the `modules` section.

**Reply by emcdonald** [reply]
My current config in ~/.config/nvim/lua/plugins/treesitter.lua


```lua
return {
  {
    "nvim-treesitter/nvim-treesitter",
    ---@type TSConfig
    opts = {
      prefer_git = true,
    },
  },
}
```

**Reply by dpetka2001** [reply]
Yes you are right. It's inside the `modules` section. I don't know if that's the correct place that it gets put, but the configuration above is the correct way to do it i believe. Is there any kind of way to test if it still uses curl except from `TSConfigInfo`? Maybe ask at treesitter repo if there's some way to debug if the option is indeed set?

**Comment by folke** [comment, marked as answer]
LazyVim opts for treesitter are not passed to the configs module, not the install module. See here https://github.com/LazyVim/LazyVim/blob/cbf1d335ed6a478a2e6144aa2d462a8330b2b0fc/lua/lazyvim/plugins/treesitter.lua#L81

You can do something like the below:

```lua
{"treesitter..", opts = function(_, opts)
  require("nvim-treesitter.install").prefer_git = true
end}
```

**Reply by dpetka2001** [reply]
As always thank you for your help, even though I'm not the OP, it's useful to know how it works. I thought the `configs` and `install` modules were just commands exposed to the `setup` function of treesitter and thought since `config` and `opts` of Lazyvim both call the setup function of the plugins it would be ok.

**Reply by emcdonald** [reply]
Thank you very much. That solved my issue. Much appreciated.

**Comment by emcdonald** [comment]
For anyone interested, here is the complete config based on @folke's answer.

`~/.config/nvim/lua/plugins/treesitter.lua`

```lua
return {
  {
    "nvim-treesitter/nvim-treesitter",
    opts = function(_, opts)
      require("nvim-treesitter.install").prefer_git = true
    end,
  },
}
```

**Reply by evadeflow** [reply]
This was a big help to me just now, thanks. Thought I'd 'pay it forward' by mentioning that you can set *multiple* install options by assigning the install module to a temporary variable, like so:

```lua

  {
    -- Highlight, edit, and navigate code
    'nvim-treesitter/nvim-treesitter',
    opts = function(_, opts)
      local tsi = require("nvim-treesitter.install")
      tsi.prefer_git = false
      tsi.compilers = { "clang", "gcc" }
    end,
    dependencies = {
      'nvim-treesitter/nvim-treesitter-textobjects',
    },
    build = ':TSUpdate',
  },
```

I just did this to get neovim working natively on my Windows 10 work laptop with a bare-bones `init.lua` from  `kickstarter.nvim` . The [nvim-treesitter wiki](https://github.com/nvim-treesitter/nvim-treesitter/wiki/Windows-support) recommended setting `prefer_git = false`. And there's a `cc` utility in my `PATH` that belongs to the ClearCase SCM 🤮, so I needed to change the `compilers` search order.  Installing Chocolatey and running these commands got me a mostly working environment, with the pywright and clangd LSPs:

```
choco install --pre neovim
choco install llvm
choco install unzip
choco install wget
choco install gzip
choco install pwsh
choco install ripgrep
choco install fd
choco install nodejs
```

I'm still futzing with it, but I'm *really* happy with how it's shaping up so far. Couldn't have got it working without the hint from @folke and you, so... *thank you* guys!
